tools/state-viewer/src/apply_chain_range.rs

```diff
@@ -286,43 +286,54 @@ fn apply_block_from_range(
-                    hash: block_hash,
-                    height: block.header().height(),
-                    prev_hash: *block.header().prev_hash(),
+    progress_reporter.inc_and_report_progress(height, apply_result.total_gas_burnt);
+
+    match (mode, storage) {
+        (ApplyRangeMode::Benchmark, _) => {}
+        (_, StorageSource::Trie | StorageSource::TrieFree) => {}
+        (_, StorageSource::FlatStorage | StorageSource::Memtrie) => {
+            // Compute delta and immediately apply to flat storage.
+            let changes =
+                FlatStateChanges::from_state_changes(apply_result.trie_changes.state_changes());
+            let delta = near_store::flat::FlatStateDelta {
+                metadata: near_store::flat::FlatStateDeltaMetadata {
+                    block: BlockInfo {
+                        hash: block_hash,
+                        height: block.header().height(),
+                        prev_hash: *block.header().prev_hash(),
+                    },
+                    prev_block_with_changes: None,
                 },
-                prev_block_with_changes: None,
-            },
-            changes,
-        };
-
-        let flat_storage_manager = runtime_adapter.get_flat_storage_manager();
-        let flat_storage = flat_storage_manager.get_flat_storage_for_shard(shard_uid).unwrap();
-        let store_update = flat_storage.add_delta(delta).unwrap();
-        store_update.commit().unwrap();
-        flat_storage.update_flat_head(&block_hash).unwrap();
+                changes,
+            };
 
-        // Apply trie changes to trie node caches.
-        let mut fake_store_update = read_store.trie_store().store_update();
-        apply_result.trie_changes.insertions_into(&mut fake_store_update);
-        apply_result.trie_changes.deletions_into(&mut fake_store_update);
-    } else {
-        if let Err(err) = maybe_save_trie_changes(
-            write_store,
-            genesis.config.genesis_height,
-            apply_result,
-            height,
-            shard_id,
-        ) {
-            panic!("Error while saving trie changes at height {height}, shard {shard_id} ({err})");
+            let flat_storage_manager = runtime_adapter.get_flat_storage_manager();
+            let flat_storage = flat_storage_manager.get_flat_storage_for_shard(shard_uid).unwrap();
+            let store_update = flat_storage.add_delta(delta).unwrap();
+            store_update.commit().unwrap();
+            flat_storage.update_flat_head(&block_hash).unwrap();
+        }
+    }
+    match (mode, storage) {
+        (ApplyRangeMode::Benchmark, _) => {}
+        (_, StorageSource::FlatStorage) => {
+            // Apply trie changes to trie node caches.
+            let mut fake_store_update = read_store.trie_store().store_update();
+            apply_result.trie_changes.insertions_into(&mut fake_store_update);
+            apply_result.trie_changes.deletions_into(&mut fake_store_update);
```

Comment:
Out of curiosity what is fake about this store update and why use a fake one? Is it fake because it's the read store? It all seems a bit wild though I guess it's fine for tooling, and it was here before your PR anyway.